proto/docs/p4runtime_type_serialization.md

```diff
@@ -0,0 +1,96 @@
+# Arbitary P4_16 type serialization in P4Runtime
+
+## Problem statement
+The P4_16 language includes more complex types than just bitstrings or even
+headers (see [P4_16
+specification](https://p4.org/p4-spec/docs/P4-16-v1.0.0-spec.html#sec-p4-type)). Most
+of these complex data types can be exposed to the control-plane through table
+key expressions, value set lookup expressions, register (PSA extern type) value
+types, etc. Not supporting these more compex types can be very limiting. For
+example, the following P4_16 objects involve complex types that need to be
+exposed in P4Runtime in order to support runtime operations on these objects.
+
+```
+value_set<tuple<bit<16>, bit<8> > > pvs_complex;
+state parse_ipv4 {
+   packet.extract<ipv4_t>(hdr.ipv4);
+   transition select({ hdr.ipv4.version, hdr.ipv4.protocol }){
+       pvs_complex: parse_inner;
+       default: accept;
+   }
+}
+```
+
+```
+header_union ip_t {
+   ipv4_t ipv4;
+   ipv6_t ipv6;
+}
+Register<ip_t, bit<32> >(128) register_ip;
+```
+
+One solution would be to use only bitstrings (`bytes` type) in p4runtime.proto
+and to define a custom serialization format for complex P4_16 types. The
+serialization would maybe be trivial for header types but would require some
+work for header unions, stacks, etc. Rather than coming-up with a serialization
+format from scratch, we decided to use a Protobuf representation for all P4_16
+types, since the P4Runtime service already uses Protobuf messages.
+
+## P4 type specifications in p4info.proto
+In order for the P4Runtime client to generate correctly-formatted messages and
+for the P4Runtime service implementation to validate them, P4Info needs to
+specify the type of each P4 expression which is exposed to the control-plane. In
+the register example above, client and server need to know that each element of
+the register has type `ip_t`, which is a header union with 2 possible headers:
+`ipv4` with type `ipv4_t` and `ipv6` with type `ipv6`. Similarly, they need to
+know the field layout of both these header types.
+
+To achieve this we introduce 2 main protobuf messages: `P4TypeInfo` and
+`P4DataTypeSpec`.
+
+`P4TypeInfo` is a top-level member of `P4Info` and includes Protobuf maps
+storing the type specification for all the named types in the P4_16
+program. These named types are `struct`, `header`, `header_union` and `enum`;
+for each of these we have a type specification message, respectively
+`P4StructTypeSpec`, `P4HeaderTypeSpec`, `P4HeaderUnionTypeSpec` and
+`P4EnumTypeSpec`. We support P4 annotations for named types, which is useful to
```

Comment:
I wonder whether you will care about serializing match_kind ever.
It's just a special enum.

Reply:
I cannot think of a case where that would be useful to the control plane. From a language perspective, can I even instantiate a variable of type `match_kind`?